File: data-azure-cosmos/src/test/java/io/micronaut/data/azure/repositories/FamilyRepository.java
package io.micronaut.data.azure.repositories;

import com.azure.cosmos.models.PartitionKey;
import io.micronaut.context.annotation.Parameter;
import io.micronaut.core.annotation.NonNull;
import io.micronaut.core.annotation.Nullable;
import io.micronaut.data.annotation.Id;
import io.micronaut.data.annotation.Join;
import io.micronaut.data.annotation.Query;
import io.micronaut.data.azure.entities.Address;
import io.micronaut.data.azure.entities.Child;
import io.micronaut.data.azure.entities.Family;
import io.micronaut.data.azure.entities.GenderAware;
import io.micronaut.data.azure.entities.PetType;
import io.micronaut.data.cosmos.annotation.CosmosRepository;
import io.micronaut.data.model.jpa.criteria.PersistentEntityCriteriaBuilder;
import io.micronaut.data.repository.PageableRepository;
import io.micronaut.data.repository.jpa.JpaSpecificationExecutor;
import io.micronaut.data.repository.jpa.criteria.PredicateSpecification;

import java.util.Arrays;
import java.util.Date;
import java.util.List;
import java.util.Map;
import java.util.Optional;

@CosmosRepository
public abstract class FamilyRepository implements PageableRepository<Family, String>, JpaSpecificationExecutor<Family> {

    @Override
    @NonNull
    public abstract Optional<Family> findById(@NonNull String id);

    public abstract void updateRegistered(@Id String id, boolean registered);

    public abstract void updateRegistered(@Id String id, boolean registered, PartitionKey partitionKey);

    public abstract void updateAddress(@Parameter("id") @Id String id, @NonNull @Parameter("address") Address address);

    // Raw query for Cosmos update is not supported and calling this method will throw an error.
    @Query("UPDATE family f SET f.lastName=@p1 WHERE f.id=@p2")
    public abstract void updateLastName(String id, String lastName);

    public abstract void deleteByLastName(String lastName, PartitionKey partitionKey);

    public abstract void deleteById(String id, PartitionKey partitionKey);

    // Raw query not supported for delete so this would throw an error
    @Query("DELETE FROM family f WHERE f.registered=@p1")
    public abstract void deleteByRegistered(boolean registered);

    public abstract boolean existsByIdAndRegistered(String id, boolean registered);

    public abstract long countByRegistered(boolean registered);

    @Query("SELECT VALUE f.lastName FROM family f ORDER BY f.lastName DESC OFFSET 0 LIMIT 1")
    public abstract String lastOrderedLastName();

    @Query("SELECT VALUE f.registeredDate FROM family f WHERE NOT IS_NULL(f.registeredDate) ORDER BY f.registeredDate DESC OFFSET 0 LIMIT 1")
    public abstract Date lastOrderedRegisteredDate();

    public abstract List<Family> findByAddressStateAndAddressCityOrderByAddressCity(String state, String city);

    public abstract void updateByAddressCounty(String county, boolean registered, @Nullable Date registeredDate);

    @Join("children.pets")
    public abstract List<Family> findByChildrenPetsType(PetType type);

    public abstract List<Child> findChildrenByChildrenPetsGivenName(String name);

    public abstract List<Family> findByIdIn(List<String> ids);

    public abstract List<Family> findByIdNotIn(List<String> ids);

    public abstract List<Family> findByLastNameLike(String lastName);

    @Query("SELECT DISTINCT VALUE f FROM family f JOIN c IN f.children WHERE c.firstName IN (:firstNames)")
    public abstract List<Family> findByChildrenFirstNameIn(List<String> firstNames);

    @Query("SELECT DISTINCT VALUE f FROM family f WHERE ARRAY_CONTAINS(f.children, :gender, true)")
    public abstract List<Family> childrenArrayContainsGender(Map.Entry<String, Object> gender);

    static class Specifications {

        public static PredicateSpecification<Family> lastNameEquals(String lastName) {
            return (root, criteriaBuilder) -> criteriaBuilder.equal(root.get("lastName"), lastName);
        }

        public static PredicateSpecification<Family> idsIn(String... ids) {
            return (root, criteriaBuilder) -> root.get("id").in(Arrays.asList(ids));
        }

        public static PredicateSpecification<Family> idsNotIn(String... ids) {
            return (root, criteriaBuilder) -> root.get("id").in(Arrays.asList(ids)).not();
        }

        public static PredicateSpecification<Family> idsInAndNotIn(List<String> idsIn, List<String> idsNotIn) {
            return (root, criteriaBuilder) -> criteriaBuilder.and(root.get("id").in(idsIn), root.get("id").in(idsNotIn).not());
        }

        public static PredicateSpecification<Family> registeredEquals(boolean registered) {
            return (root, criteriaBuilder) -> criteriaBuilder.equal(root.get("registered"), registered);
        }

        public static PredicateSpecification<Family> childrenArrayContainsGender(GenderAware gender) {
            return (root, criteriaBuilder) -> ((PersistentEntityCriteriaBuilder) criteriaBuilder).arrayContains(root.join("children"), criteriaBuilder.literal(gender));
        }
    }
}

File: doc-examples/azure-cosmos-example-python/src/main/python/example/FamilyRepository.py
from typing import Annotated

from com.azure.cosmos.models import PartitionKey
from java.util import Date, Map, Optional
from micronaut.context.annotation import Parameter
from micronaut.data.annotation import Id, Join, Query
from micronaut.data.cosmos.annotation import CosmosRepository
from micronaut.data.repository import PageableRepository

from example.Address import Address
from example.Child import Child
from example.Family import Family
from example.GenderAware import GenderAware
from example.Pet import PetType


# TODO(python): the specification methods are declared on the repository instead of extending JpaSpecificationExecutor:
# its same-arity overloads (PredicateSpecification / QuerySpecification) are ambiguous for a Python lambda, see DISABLED_TESTS.md
@CosmosRepository
class FamilyRepository(PageableRepository[Family, str]):

    def updateRegistered(self, id: Annotated[str, Id], registered: bool) -> None: ...

    def updateAddress(self, id: Annotated[str, Id, Parameter("id")], address: Annotated[Address, Parameter("address")]) -> None: ...

    # Raw query for Cosmos update is not supported and calling this method will throw an error.
    @Query("UPDATE family f SET f.lastName=@p1 WHERE f.id=@p2")
    def updateLastName(self, id: str, lastName: str) -> None: ...

    # tag::partitionkey[]
    def queryById(self, id: str, partitionKey: PartitionKey) -> Optional[Family]: ...

    def updateRegisteredById(self, id: Annotated[str, Id], registered: bool, partitionKey: PartitionKey) -> None: ...

    def deleteByLastName(self, lastName: str, partitionKey: PartitionKey) -> None: ...

    def removeById(self, id: str, partitionKey: PartitionKey) -> None: ...
    # end::partitionkey[]

    # Raw query not supported for delete so this would throw an error
    @Query("DELETE FROM family f WHERE f.registered=@p1")
    def deleteByRegistered(self, registered: bool) -> None: ...

    def existsByIdAndRegistered(self, id: str, registered: bool) -> bool: ...

    def countByRegistered(self, registered: bool) -> int: ...

    @Query("SELECT VALUE f.lastName FROM family f ORDER BY f.lastName DESC OFFSET 0 LIMIT 1")
    def lastOrderedLastName(self) -> str: ...

    @Query("SELECT VALUE f.registeredDate FROM family f WHERE NOT IS_NULL(f.registeredDate) ORDER BY f.registeredDate DESC OFFSET 0 LIMIT 1")
    def lastOrderedRegisteredDate(self) -> Date: ...

    # tag::relations[]
    def findByAddressStateAndAddressCityOrderByAddressCity(self, state: str, city: str) -> list[Family]: ...

    def updateByAddressCounty(self, county: str, registered: bool, registeredDate: Date | None) -> None: ...

    @Join(value="children.pets", alias="pets")
    def findByChildrenPetsType(self, type: PetType) -> list[Family]: ...

    def findChildrenByChildrenPetsGivenName(self, name: str) -> list[Child]: ...
    # end::relations[]

    def findByIdIn(self, ids: list[str]) -> list[Family]: ...

    def findByIdNotIn(self, ids: list[str]) -> list[Family]: ...

    def findByLastNameLike(self, lastName: str) -> list[Family]: ...

    # tag::method_array_contains[]
    def findByTagsArrayContains(self, tag: str) -> list[Family]: ...
    # end::method_array_contains[]

    # tag::array_contains_partial[]
    @Query("SELECT DISTINCT VALUE f FROM family f WHERE ARRAY_CONTAINS(f.children, :gender, true)")
    def childrenArrayContainsGender(self, gender: Map.Entry[str, object]) -> list[Family]: ...
    # end::array_contains_partial[]


def last_name_equals(last_name: str):
    return lambda root, criteria_builder: criteria_builder.equal(root.get("lastName"), last_name)


def ids_in(*ids: str):
    return lambda root, criteria_builder: root.get("id").in_(list(ids))


def ids_not_in(*ids: str):
    return lambda root, criteria_builder: root.get("id").in_(list(ids)).not_()


# tag::predicate_array_contains[]
def tags_contain(tag: str):
    return lambda root, criteria_builder: criteria_builder.arrayContains(root.get("tags"), criteria_builder.literal(tag))
# end::predicate_array_contains[]


# tag::predicate_array_contains_partial[]
def children_array_contains_gender(gender: GenderAware):
    return lambda root, criteria_builder: criteria_builder.arrayContains(root.join("children"), criteria_builder.literal(gender))
# end::predicate_array_contains_partial[]
